sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/metrics/MetricsReader.java

```diff
@@ -0,0 +1,126 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.testutils.metrics;
+
+import java.util.NoSuchElementException;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.metrics.DistributionResult;
+import org.apache.beam.sdk.metrics.MetricNameFilter;
+import org.apache.beam.sdk.metrics.MetricQueryResults;
+import org.apache.beam.sdk.metrics.MetricResult;
+import org.apache.beam.sdk.metrics.MetricsFilter;
+import org.joda.time.Duration;
+import org.slf4j.LoggerFactory;
+
+/** Provides methods for querying metrics from {@link PipelineResult} per namespace. */
+public class MetricsReader {
+
+  private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(MetricsReader.class);
+
+  private enum DistributionType {
+    MIN,
+    MAX
+  }
+
+  private final PipelineResult result;
+
+  private final String namespace;
+
+  public MetricsReader(PipelineResult result, String namespace) {
+    this.result = result;
+    this.namespace = namespace;
+  }
+
+  /**
+   * Return the current value for a long counter, or a default value if can't be retrieved. Note
+   * this uses only attempted metrics because some runners don't support committed metrics.
+   */
+  public long getCounterMetric(String name, long defaultValue) {
+    MetricQueryResults metrics =
+        result
+            .metrics()
+            .queryMetrics(
+                MetricsFilter.builder()
+                    .addNameFilter(MetricNameFilter.named(namespace, name))
+                    .build());
+    Iterable<MetricResult<Long>> counters = metrics.getCounters();
+    try {
+      MetricResult<Long> metricResult = counters.iterator().next();
+      return metricResult.getAttempted();
+    } catch (NoSuchElementException e) {
+      LOG.error("Failed to get metric {}, from namespace {}", name, namespace);
+    }
+    return defaultValue;
+  }
+
+  /**
+   * Return start time metric by counting the difference between "now" and min value from a
+   * distribution metric.
+   */
+  public long getStartTimeMetric(long now, String name) {
+    return this.getTimestampMetric(now, this.getDistributionMetric(name, DistributionType.MIN, -1));
+  }
+
+  /**
+   * Return end time metric by counting the difference between "now" and MAX value from a
+   * distribution metric.
+   */
+  public long getEndTimeMetric(long now, String name) {
+    return this.getTimestampMetric(now, this.getDistributionMetric(name, DistributionType.MAX, -1));
+  }
+
+  /**
+   * Return the current value for a long counter, or a default value if can't be retrieved. Note
+   * this uses only attempted metrics because some runners don't support committed metrics.
+   */
+  private long getDistributionMetric(String name, DistributionType distType, long defaultValue) {
```

Comment:
This default value is allways -1. Maybye I should remove this parameter and always set -1 in the method body?

Reply:
In this case I agree. It would help avoid "magic" -1 in tests code. Maybe it's worth introduce it as a global variable since there is magic `-1` in `getTimestampMetric` as well (if it's gonna stay there).